source/heating_model/shear_heating.cc

```diff
@@ -41,15 +41,30 @@ namespace aspect
 
       for (unsigned int q=0; q<heating_model_outputs.heating_source_terms.size(); ++q)
         {
-          const SymmetricTensor<2,dim> compressible_strain_rate
-            = (this->get_material_model().is_compressible()
-               ?
-               material_model_inputs.strain_rate[q] - 1./3. * trace(material_model_inputs.strain_rate[q]) * unit_symmetric_tensor<dim>()
-               :
-               material_model_inputs.strain_rate[q]);
-
-          heating_model_outputs.heating_source_terms[q] = 2.0 * material_model_outputs.viscosities[q] *
-                                                          compressible_strain_rate * compressible_strain_rate;
+          const SymmetricTensor<4,dim> C = material_model_outputs.constitutive_tensors[q];
+          const SymmetricTensor<2,dim> directed_strain_rate = ((C == dealii::identity_tensor<dim> ())
```

Comment:
This is a potentially expensive comparison: it requires comparing all 36 elements of the two tensors.

If you assume that the elements of the constitutive_tensors array are set to anything at all (which you do here), then I'd simply always do the multiplication. Loading the data from memory so you can do the comparison will in practice not take more time than actually doing the multiplication. But it will keep the code easier to read.